plugins/module_utils/identity/keycloak/keycloak.py

```diff
@@ -280,6 +311,45 @@ def __init__(self, module, connection_header):
         self.restheaders = connection_header
         self.http_agent = self.module.params.get('http_agent')
 
+    def _request(self, url, method, data=None):
+        def make_request_ignoring_401():
+            try:
+                return open_url(url, method=method, data=data,
+                                http_agent=self.http_agent, headers=self.restheaders,
+                                timeout=self.connection_timeout,
+                                validate_certs=self.validate_certs)
+            except HTTPError as e:
+                if e.code != 401:
+                    raise e
+
+            return None
+
+        r = make_request_ignoring_401()
+        if r is not None:
+            return r
+
+        # Authentication may have expired, re-authenticate with refresh token and retry
+        refresh_token = self.module.params.get('refresh_token')
+        if refresh_token is not None:
+            token = _get_token_using_refresh_token(self.module.params)
+            self.restheaders['Authorization'] = 'Bearer ' + token
+
+        r = make_request_ignoring_401()
+        if r is not None:
+            return r
```

Comment:
I think we should not make a new request if the access token has not been refreshed.
```suggestion
        if refresh_token is not None:
            token = _get_token_using_refresh_token(self.module.params)
            self.restheaders['Authorization'] = 'Bearer ' + token

            r = make_request_ignoring_401()
            if r is not None:
                return r
```

Reply:
Thanks, good catch; moved the re-requests inside the if statements to avoid making them unnecessarily.